### PR TITLE
Protect: Parse data from the initial state

### PR DESCRIPTION
**Pull request by leogermani**
This PR adds a custom hook that reads and parses data from the initial state and make it available to be used by components.

It's a hook because it started out using `useContext()` but then I realized it was not needed. Even though it is not technically a hook, it looks good. But we could refactor it as a simple export/import if you prefer.

#### Changes proposed in this Pull Request:

* Parse data from initial state and use it in the UI

#### Jetpack product discussion

N/A

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

Make sure to add to your wp-config:
```
define( 'JETPACK_PROTECT_DEV__BYPASS_CACHE', true );
define( 'JETPACK_PROTECT_DEV__API_RESPONSE_TYPE', 'complete_green' );
```

Visit Protect and confirm the information matches your list of installed plugins and themes.

Play with the response_type to see different results. Possible values are:
* complete_green
* complete
* incomplete_green
* incomplete
* empty -> will return the status of when the first check was not performed yet

complete means all installed plugins were checked
incomplete means some installed plugins were not checked

### CODE REVIEW

**Review by retrofox — COMMENTED** [review]
My first feedback is asking why we add the _context_ word to the hook name. Context sounds like a reserved word to me in the React context (:smart:)

**Comment by leogermani** [comment]
> My first feedback is asking why we add the _context_ word to the hook name. Context sounds like a reserved word to me in the React context (:smart:)

right! Because I started developing it using Context and changed it. I'll rename it

**Comment by retrofox** [comment]
When the site is not connected, the status contains the respective error:

image

That produce a breaking error:

image

File: projects/plugins/protect/src/js/hooks/use-protect-context/index.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ * @param {object} installed - The list of installed extensions, where each attribute key is the extension slug.
+ * @param {object} checked   - The list of checked extensions.
+ * @returns {Array} Normalized list of extensions.
+ */
+function mergeInstalledAndCheckedLists( installed, checked ) {
+	const newList = [];
+	for ( const slug in installed ) {
+		if ( checked.hasOwnProperty( slug ) ) {
+			// camelCase fixedIn.
```
Comment (retrofox) [inline projects/plugins/protect/src/js/hooks/use-protect-context/index.js:14]: Is it something that we need to adjust because of the response body shape?

Comment (leogermani) [reply projects/plugins/protect/src/js/hooks/use-protect-context/index.js:14]: Yes.

On one side, we have the list of plugins currently installed, with name, description and everything.

On the other side, we have the response from the server, with the list of plugins that were checked. In this list we only have the plugin slug.

So there might be a plugin that was recently installed and, therefore, was not checked yet. It must be in the list.

And for the plugins that were checked, we need to fetch its name from the list of installed plugins.

**Review by retrofox — COMMENTED** [review]
I get the idea about implementing this custom hook. AFAIK, and so far, this hook doesn't seem to implement or use any state logic needed for the components.
In other words, it probably could be implemented with a regular function.

However, using a custom hook I think it's ok, also keeping in mind that probably these data could get updated asynchronously?

File: projects/plugins/protect/src/js/hooks/use-protect-context/index.js (hunk 1 above)
Comment (retrofox) [reply projects/plugins/protect/src/js/hooks/use-protect-context/index.js:14]: Nice, thanks. I also mean specifically to the camel case issue.

Comment (leogermani) [reply projects/plugins/protect/src/js/hooks/use-protect-context/index.js:14]: Yes, the data comes from the server in snake-case... This is something we will most likely have to adjust once the real API is developed so I figured it's ok to do that for now. What we need to make consistent is the names and the data structure from this hook to our components.

**Comment by leogermani** [comment]
> I get the idea about implementing this custom hook. AFAIK, and so far, this hook doesn't seem to implement or use any state logic needed for the components. In other words, it probably could be implemented with a regular function.
> 
> However, using a custom hook I think it's ok, also keeping in mind that probably these data could get updated asynchronously?

Yeah, as I said in the description, this could be a regular function, but a hook does not look so bad and leaves the door open to improvements in the future. But I could change that if you think it's better to use a function.

I've renamed the hook and handled the error when the site is not connected

**Comment by retrofox** [comment]
> > I get the idea about implementing this custom hook. AFAIK, and so far, this hook doesn't seem to implement or use any state logic needed for the components. In other words, it probably could be implemented with a regular function.
> > However, using a custom hook I think it's ok, also keeping in mind that probably these data could get updated asynchronously?
> 
> ... but a hook does not look so bad and leaves the door open to improvements in the future. But I could change that if you think it's better to use a function.

:deal:

**Review by retrofox — COMMENTED** [review]
I still think we need to handle the breaking bug when the user is connected


https://user-images.githubusercontent.com/77539/164460822-01eebe78-9700-479b-8024-3e79e1e30809.mov

File: projects/plugins/protect/src/js/hooks/use-protect-data/index.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	if ( wpVersion && status.wordpress && status.wordpress.version === wpVersion ) {
+		core = status.wordpress;
+		core.name = 'wp';
+	} else {
+		core = {
+			version: wpVersion,
+			vulnerabilities: [],
+			name: 'wp',
+		};
+	}
```
Comment (renatoagds) [inline projects/plugins/protect/src/js/hooks/use-protect-data/index.js:67]: Vote to move this out to some utility function as you did with `mergeInstalledAndCheckedLists`

**Comment by retrofox** [comment]
You're going to hate me but I think it should be split out in two PRs

**Review by renatoagds — APPROVED** [review]
LGTM

**Comment by leogermani** [comment]
r244162-wpcom

**Pull request merged**